### ISSUE TITLE
False positive for interpolated selector containing a colon in declaration-colon-space-after and declaration-block-trailing-semicolon

### ISSUE THREAD

**Issue by DanielaValero** [stylelint/stylelint#1936]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

After updating to the version 7.3.1, stylelint is failing on the rules: 
 * declaration-colon-space-after and 
 * declaration-block-trailing-semicolon 

In a Sass file doing:

```
  &:#{$position} {
```

Stylelint now believes that this is a CSS property.

> Which rule, if any, is this issue related to?

 * declaration-colon-space-after and 
 * declaration-block-trailing-semicolon 

> What CSS is needed to reproduce this issue?

```css
@mixin supergraphic-line($position: before, $height: $supergraphic-height, $width: 100%) {

  &:#{$position} {
    position: absolute;
    left: 0;
    width: $width;
    height: $height;
    content: '';
    background-repeat: no-repeat;
    background-size: cover;
    background-position: center center;
    background-image: url($supergraphic-path--xs);
  }
}
```

> What stylelint configuration is needed to reproduce this issue?

```js
/**
 * Default configuration to lint
 * the airbnb css style-guide.
 * This file is taken from a pull request to Airbnb/css repo
 * which intends to create a default preset that can be used
 * in the future. Until it is merged, we will have the config here.
 * https://github.com/airbnb/css/pull/23
 * Add more rules: http://stylelint.io/user-guide/rules/
 * Also, to understand better who the rules are named:
 * http://stylelint.io/user-guide/about-rules/
 * For an extended sample of config, check this out
 * http://stylelint.io/user-guide/example-config/
 */

function airBnbSelector(componentName) {
  const word = '[a-z0-9][a-zA-Z0-9]*';
  const descendant = `(?:__${word})?`;
  const modifier = `(?:--${word}(?:\\.${componentName}${descendant}--${word})*)?`;
  const attribute = '(?:\\[.+\\])?';
  const state = `(?:\\.is-${word})*`;
  const body = descendant +
    modifier +
    attribute +
    state;
  return new RegExp(`^\\.${componentName}\\b${body}$`);
}

export default {
  plugins: [
    'stylelint-scss',
    'stylelint-selector-bem-pattern',
  ],
  rules: {

    /* ==========================================================================
       Airbnb config
       ========================================================================== */

    // CSS formatting
    'selector-no-id': true,
    indentation: 2,
    'selector-list-comma-newline-after': 'always',
    'declaration-colon-space-after': 'always',
    'declaration-colon-space-before': 'never',
    'block-opening-brace-space-before': 'always',
    'declaration-block-single-line-max-declarations': 1,
    'rule-nested-empty-line-before': ['always', {
      ignore: ['after-comment'],
    },
  ],
    'rule-non-nested-empty-line-before': ['always', {
      ignore: ['after-comment'],
    },
  ],

    // Comments
    'comment-empty-line-before': ['always', {
      ignore: ['stylelint-commands'],
    },
  ],

    // Bem
    'plugin/selector-bem-pattern': {
      componentName: /^[A-Z][a-zA-Z0-9]+$/,
      componentSelectors: airBnbSelector,
      utilitySelectors: /^\.u(?:-[a-z0-9][a-zA-Z0-9]*)+$/,
    },

    // Border
    'declaration-property-value-blacklist': {
      '/^border/': ['none'],
    },

    // SASS
    // All @includes after properties
    // Nested selectors after properties

    // constiables dash-dashed
    // This regexp matches:
    // $button-text-background-color--hover-hola
    // regex under construction
    // 'scss/dollar-constiable-pattern': '\b[a-z]+(?:-)+(\b[a-z]+(?:-))*',

    // forbid extend
    'at-rule-blacklist': ['extend'],

    // Nesting depth
    'max-nesting-depth': 3,

    /* ==========================================================================
       Best practices
       ========================================================================== */

    // Specificity
    // To learn more about this:
    // http://csswizardry.com/2014/07/hacks-for-dealing-with-specificity/
    // "id,class,type",
    // selector-max-specificity
    'declaration-no-important': true,
    'selector-max-compound-selectors': 3,
    'selector-no-qualifying-type': true,

    // Selectors
    'no-duplicate-selectors': true,

    // Blocks
    'block-no-empty': true,
    'at-rule-empty-line-before': [
      'always', {

        // Allow mixins to have an empty line before
        ignoreAtRules: ['import', 'first-nested', 'else'],
      },
    ],

    // More styling rules for more consistency
    'at-rule-name-case': 'lower',

    // Colors
    'color-hex-case': 'lower',
    'color-hex-length': 'long',
    'color-no-invalid-hex': true,
    'no-indistinguishable-colors': [true, {
      ignore: ['#000000', '#dddddd', '#ffffff', '#808080'],
      threshold: 1,
    },
  ],

    // strings
    'string-quotes': 'single',

    // Values
    // Disallow vendor prefix, they are added by autoprefixer
    'value-no-vendor-prefix': true,
    'value-list-comma-space-after': 'always-single-line',

    // Disallows margin: 1px 1px 1px 1px;
    'shorthand-property-no-redundant-values': true,

    // Comments
    'comment-whitespace-inside': 'always',

    // Functions
    'function-comma-space-after': 'always-single-line',
    'function-comma-space-before': 'never',

    // Numbers
    // unitless zero and no trailing zeros
    'length-zero-no-unit': true,
    'number-no-trailing-zeros': true,

    // Syntax
    'declaration-block-trailing-semicolon': 'always',

    // Declaration blocks
    'declaration-block-no-duplicate-properties': true,

    // More on this one here:
    // http://stylelint.io/user-guide/rules/declaration-block-no-ignored-properties/
    'declaration-block-no-ignored-properties': true,

    // Require a reason before or after a stylelint disable
    'stylelint-disable-reason': 'always-after',

    'scss/at-extend-no-missing-placeholder': true,

    // Prevents adding unnecesary Specificity or complicated sass stuff
    'scss/selector-no-redundant-nesting-selector': true,

  },
};
```

> Which version of stylelint are you using?

7.3.1

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Pure stylelint

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

yes, scss.

> What did you expect to happen?

That it does not fail, as it did in the previous version.

> What actually happened (e.g. what warnings or errors you are getting)?

I got an error
```shell

app/core/styles/mixins/supergraphic-line.scss
22:5	⚠  Expected single space after ":" (declaration-colon-space-after) [stylelint]
22:16	⚠  Expected a trailing semicolon (declaration-block-trailing-semicolon) [stylelint]
37:5	⚠  Expected single space after ":" (declaration-colon-space-after) [stylelint]
37:16	⚠  Expected a trailing semicolon (declaration-block-trailing-semicolon) [stylelint]

```

**Comment by jeddy3** [comment]
@DanielaValero Thank you for the reporting the issue and for using the template.

I can reproduce the problem with the following reduced code:

```scss
@mixin name($var) {
  &:#{$var} {}
}
```

```json
{
  "rules": {
     "declaration-colon-space-after": "always",
     "declaration-block-trailing-semicolon": "always",
  }
}
```

Gives:

```
 2:5   ✖  Expected single space after ":"   declaration-colon-space-after
 2:16  ✖  Expected a trailing semicolon     declaration-block-trailing-semicolon
```

I will label this as a bug.

> That it does not fail, as it did in the previous version.

Do you know what previous version you were using? 

This issue might be caused by a change in the behaviour of [`postcss-scss` parser](https://github.com/postcss/postcss-scss) at some point. @DanielaValero If you'd like to help out you could investigate if there have been any recent changes to [`postcss-scss` parser] that might explain how `&:#{$var} {}` is being parsed differently now.

**Comment by DanielaValero** [comment]
Hello @jeddy3 

The version we had before was the 7.0.1.

I have done a mini research and realized that indeed this bug is introduced in the version 0.2.0 of postscss-scss when they introduced support for nested properties.

I don't understand fully how postcss-scss is doing the magic behind, however I tend to think that it might be that the option of having this syntax was not added yet perhaps?

**Comment by jeddy3** [comment]
> I have done a mini research and realized that indeed this bug is introduced in the version 0.2.0 of postscss-scss when they introduced support for nested properties.

Thanks for tracking that down!

Would you be able to raise an issue in the `postcss-scss` issue tracker about this problem?

**Comment by DanielaValero** [comment]
Hi @jeddy3 

Yes, I have just opened it: https://github.com/postcss/postcss-scss/issues/45

I would say that we could close this report as it is related to postscss-scss and not to stylelint.

Greetings

Daniela

**Comment by ai** [comment]
Fix released in `postcss-scss` 0.3.1 release

**Comment by jeddy3** [comment]
@ai Perfect, `0.3.1` works a treat! Thanks, as always, for the very speedy fix.

@DanielaValero Thanks for following through with the issue to the `postcss-scss` tracker. 

Run `npm install stylelint@latest` to pull down the `0.3.1` version of the `postcss-scss` dependency into your local `stylelint` install.
